pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +351,446 @@ pub mod pallet {
+
+						if pending_amount.is_zero() {
+							// NOTE: How make this works with market ratios?
+							let remaining_foreign_amount = T::CurrencyConverter::stable_to_stable(
+								info.base.foreign_currency,
+								Pallet::<T>::pool_currency_of(investment_id)?,
+								T::Investment::investment(&who, investment_id)?,
+							)?;
+
+							T::DecreasedForeignInvestOrderHook::notify_status_change(
+								(who.clone(), investment_id),
+								ExecutedForeignDecreaseInvest {
+									amount_decreased: info.decrease_swapped_amount,
+									foreign_currency: info.base.foreign_currency,
+									amount_remaining: remaining_foreign_amount,
+								},
+							)?;
+
+							info.decrease_swapped_amount = T::Balance::default();
+							info.decrease_swap_id = None;
+						}
+					}
+				}
+
+				Ok::<_, DispatchError>(())
+			})?;
+
+			ForeignRedemptionInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				if let Some(info) = maybe_info {
+					if info.swap_id == Some(swap_id) {
+						info.swapped_amount.ensure_add_assign(last_swap.amount_in)?;
+						if pending_amount.is_zero() {
+							let redemption = T::Investment::redemption(&who, investment_id)?;
+
+							T::CollectedForeignRedemptionHook::notify_status_change(
+								(who.clone(), investment_id),
+								ExecutedForeignCollect {
+									currency: info.base.foreign_currency,
+									amount_currency_payout: info.swapped_amount,
+									amount_tranche_tokens_payout: info.collected_tranche_tokens(),
+									amount_remaining: redemption,
+								},
+							)?;
+
+							info.pending_tranche_tokens
+								.ensure_sub_assign(info.collected_tranche_tokens())?;
+
+							info.base.collected = CollectedAmount::default();
+							info.swapped_amount = T::Balance::default();
+							info.swap_id = None;
+
+							if info.pending_tranche_tokens.is_zero() {
+								*maybe_info = None;
+							}
+						}
+					}
+				}
+
+				Ok(())
+			})
+		}
+	}
+
+	pub struct CollectedInvestmentHook<T>(PhantomData<T>);
+	impl<T: Config> StatusNotificationHook for CollectedInvestmentHook<T> {
+		type Error = DispatchError;
+		type Id = (T::AccountId, T::InvestmentId);
+		type Status = CollectedAmount<T::Balance>;
+
+		fn notify_status_change(
+			(who, investment_id): (T::AccountId, T::InvestmentId),
+			collected: CollectedAmount<T::Balance>,
+		) -> DispatchResult {
+			ForeignInvestmentInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				let info = maybe_info.as_mut().ok_or(Error::<T>::InfoNotFound)?;
+				info.base.collected.increase(&collected)?;
+
+				// NOTE: How make this works with market ratios?
+				let remaining_foreign_amount = T::CurrencyConverter::stable_to_stable(
+					info.base.foreign_currency,
+					Pallet::<T>::pool_currency_of(investment_id)?,
+					T::Investment::investment(&who, investment_id)?,
+				)?;
+
+				// NOTE: How make this works with market ratios?
+				let foreign_amount = T::CurrencyConverter::stable_to_stable(
```

Comment:
Nit: To improve readability, could we rename this to 
```suggestion
				let collected_foreign_amount = T::CurrencyConverter::stable_to_stable(
```

Reply:
Totally agree!